shellai/config.py

```diff
@@ -0,0 +1,150 @@
+
+
+class History(namedtuple("History", ["enabled", "file", "max_size"])):
+    """This class represents the [history] section of our config.toml file."""
+
+    __slots__ = ()
+
+    def __new__(
+        cls,
+        enabled: bool = True,
+        file: Path = Path("~/.local/share/shellai/shellai_history.json"),
+        max_size: int = 100,
+    ):
+        file = Path(file)
+        return super(History, cls).__new__(cls, enabled, file, max_size)
+
+
+class Backend(namedtuple("Backend", ["endpoint"])):
+    """This class represents the [backend] section of our config.toml file."""
+
+    __slots__ = ()
+
+    def __new__(
+        cls,
+        endpoint: str = "http://0.0.0.0:8080/v1/query/",
+    ):
+        return super(Backend, cls).__new__(cls, endpoint)
+
+
+class Config:
+    """Class that holds our configuration file representation.
+
+    .. note::
+        With this class, after being initialized, one can access their fields like:
+
+        >>> config = Config()
+        >>> config.output.enforce_script
+
+        The currently available top-level fields are:
+            * output = Match the `py:Output` class and their fields
+            * history = Match the `py:History` class and their fields
+            * backend = Match the `py:backend` class and their fields
+    """
+
+    def __init__(self, output: dict, history: dict, backend: dict) -> None:
+        self.output: Output = Output(**output)
+        self.history: History = History(**history)
+        self.backend: Backend = Backend(**backend)
+
+
+def _create_config_file(config_file: Path) -> None:
+    """Create a new configuration file with default values."""
+
+    logging.info(f"Creating new config file at {config_file.parent}")
+    config_file.parent.mkdir(mode=0o755, exist_ok=True)
+    base_config = Config(Output()._asdict(), History()._asdict(), Backend()._asdict())
+
+    mapping = {
+        "enforce_script": json.dumps(base_config.output.enforce_script),
+        "output_file": base_config.output.file,
+        "prompt_separator": base_config.output.prompt_separator,
+        "enabled": json.dumps(base_config.history.enabled),
+        "history_file": base_config.history.file,
+        "max_size": base_config.history.max_size,
+        "endpoint": base_config.backend.endpoint,
+    }
+    config_formatted = CONFIG_TEMPLATE.format_map(mapping)
+    config_file.write_text(config_formatted)
+
+
+def _read_config_file(config_file: Path) -> Config:
+    """Read configuration file."""
+    config_dict = {}
+    try:
+        data = config_file.read_text()
+        config_dict = tomllib.loads(data)
+    except FileNotFoundError as ex:
+        logging.error(ex)
+
+    # Convert the file paths to an instance of `Path` and normalize them.
+    config_dict["history"]["file"] = Path(config_dict["history"]["file"]).expanduser()
+    config_dict["output"]["file"] = Path(config_dict["output"]["file"]).expanduser()
```

Comment:
Note: This seems like it should be handled when data is validated and normalized rather than ad hoc inside of this function.  So, somewhere inside of the `Config` class.  Ideally, you could read the typing information from the namedtuples that define the structure of the Config class and then use those to normalize the data.  But if that's too much work for now, then I would move this transformation into Config and change the NamedTuple definitions for `History` and `Output` to use `str` for these fields.  That way at each step the data types will be the same whether they come from the default value or having been read from a file.

Reply:
Removed both calls and made a helper method to expand the user path